### PR TITLE
feat(slack): umbrella Slack UX upgrade — buttons, status, reactions, slash commands

### PR DESCRIPTION
**Pull request by Wirasm**
## Summary

- **Problem**: Slack adapter feels frozen during long workflows, has no in-thread interactivity, and surfaces no cost data. Approval gates force users to leave Slack for the web UI.
- **Why it matters**: First step toward a shared team Slack instance — these are the cheapest, highest-impact UX gaps without touching the schema or breaking the single-user model.
- **What changed**: Block Kit Approve/Reject/Cancel buttons, lifecycle reactions (🔄/✅/❌) on the triggering message, an in-thread status message edited in place as DAG nodes progress, `/archon` + `/archon-workflow` slash commands over Socket Mode, `_part i/n_` annotations on split messages, and an italic cost/token footer after assistant replies.
- **What did *not* change (scope boundary)**: No `user_id` plumbing (deferred — needs schema migration), no other adapters (Telegram/Discord/GitHub untouched), no workflow engine / event emitter changes, no DB schema changes.

### Before

```
User mentions @archon in #channel
    │
    ▼
Bot creates thread, posts replies
    │
    ▼
Workflow runs silently — thread appears frozen
    │
    ▼
Approval gate hits  ──▶  user opens Web UI to click Approve
    │
    ▼
Workflow ends, final reply posted (no cost info, no status indicator)
```

### After

```
User mentions @archon in #channel
    │
    ▼
[🔄] reaction added to user's message
    │
    ▼
Bot posts [status message] in thread:
   :arrows_counterclockwise: Workflow running
   Workflow: assist · Run: a1b2c3d4 · Elapsed: 12s
   :hourglass_flowing_sand: `plan`
   :white_circle: `apply`
   [Cancel button]
    │  (chat.update as nodes start/complete, debounced 500ms)
    ▼
Approval gate?  ──▶  [Block Kit message with Approve / Reject buttons in-thread]
                     click → operation runs → message edits to "Approved by @alice — workflow resumed"
    │
    ▼
Workflow ends:
  [🔄] removed → [✅] (or [❌])
  Status message → "Workflow completed · total cost: $0.0234"
  Direct-chat replies get appended: `_cost: $0.0234 · 12.4k tokens · stop: end_turn_`

Plus: `/archon-workflow list` / `status` / `approve <id>` etc. work natively from any channel
```

### Before

```
                 ┌──────────────────────┐
                 │ WorkflowEventEmitter │
                 └──────────┬───────────┘
                            │
              ┌─────────────┴───────────┐
              │                         │
              ▼                         ▼
   ┌────────────────────┐    (no other subscribers)
   │ WebEventBridge     │
   │ (SSE → web UI)     │
   └────────────────────┘

   Slack flow:
   SlackAdapter ──▶ messageHandler ──▶ Orchestrator ──▶ Provider
        ▲                                                  │
        └────────────── sendMessage ◀──────────────────────┘
```

### After

```
                 ┌──────────────────────┐
                 │ WorkflowEventEmitter │
                 └──────────┬───────────┘
                            │
              ┌─────────────┴───────────┐
              │                         │
              ▼                         ▼
   ┌────────────────────┐    ┌─────────────────────────┐
   │ WebEventBridge     │    │ SlackWorkflowBridge [+] │
   │ (SSE → web UI)     │    │ (reactions / status /   │
   └────────────────────┘    │  buttons via Bolt app)  │
                             └────────────┬────────────┘
                                          │ getApp().client.*
                                          ▼
   Slack flow (modified):     ┌────────────────────┐
   SlackAdapter [~] ─────────▶│ SlackAdapter app   │
     │   ↑                    │ - chat.postMessage │
     │   │                    │ - chat.update [+]  │
     │   │ sendResultFooter[+]│ - reactions.add[+] │
     │   │                    │ - app.action(...)[+]
     │   │                    │ - app.command(...)[+]
     │   │                    └────────────────────┘
     ▼   │
   messageHandler ──▶ Orchestrator [~] ──▶ Provider
                       (captures `result`
                        chunk, calls
                        sendResultFooter)
```

**Connection inventory:**

| From | To | Status | Notes |
|------|----|--------|-------|
| `WorkflowEventEmitter` | `SlackWorkflowBridge` | **new** | Bridge subscribes globally, filters Slack convs via `adapter.getTriggeringMessage` |
| `SlackWorkflowBridge` | `SlackAdapter.getApp().client.chat.*` | **new** | postMessage, update for status + approval blocks |
| `SlackWorkflowBridge` | `SlackAdapter.getApp().client.reactions.*` | **new** | reactions.add/remove on triggering message |
| `SlackWorkflowBridge` | `workflowOperations.{approveWorkflow,rejectWorkflow,abandonWorkflow}` | **new** | Button handlers call existing operations |
| `SlackWorkflowBridge` | `workflowDb.getWorkflowRun` | **new** | Reads `metadata.total_cost_usd` on terminal events |
| `SlackAdapter.start()` | `app.command('/archon'\|'/archon-workflow')` | **new** | Slash command handlers |
| `Orchestrator` | `IPlatformAdapter.sendResultFooter` | **new** (optional) | Called at end of direct-chat turns; Slack implements, others no-op |
| `server/index.ts` | `SlackWorkflowBridge` | **new** | Attaches bridge before `slack.start()` |

## Label Snapshot

- Risk: `risk: low`
- Size: `size: L`
- Scope: `adapters`
- Module: `adapters:slack`

## Change Metadata

- Change type: `feature`
- Primary scope: `adapters`

## Linked Issue

- Closes #
- Related #
- Depends on # (if stacked) — none
- Supersedes # (if replacing older PR) — none

## Validation Evidence (required)

```bash
bun run validate
# All 6 checks pass:
#   check:bundled            ✓ (36 commands, 20 workflows)
#   check:bundled-skill      ✓ (21 files)
#   type-check               ✓ all 10 packages
#   lint --max-warnings 0    ✓
#   format:check             ✓
#   test                     ✓ all packages
#
# Two new test files added (25 tests total, all passing):
#   packages/adapters/src/chat/slack/blocks.test.ts          (16 tests)
#   packages/adapters/src/chat/slack/workflow-bridge.test.ts (9 tests)
```

- Evidence provided: full `bun run validate` exit-code 0 on the branch tip.
- Intentionally skipped: end-to-end Slack workspace verification — requires the new scopes (`commands`, `reactions:write`) installed on a real Slack app. Listed under Human Verification.

## Security Impact (required)

- New permissions/capabilities? **Yes** — two new Slack bot scopes: `commands` (slash commands) and `reactions:write` (lifecycle emoji). Documented in `slack.md`.
- New external network calls? **No** — same Bolt Socket Mode connection; new API verbs only (chat.update, reactions.add, reactions.remove).
- Secrets/tokens handling changed? **No** — same `SLACK_BOT_TOKEN` / `SLACK_APP_TOKEN`.
- File system access scope changed? **No**.
- **Risk/mitigation**: Block Kit button clicks are gated by the existing `SLACK_ALLOWED_USER_IDS` whitelist inside every action handler. Unauthorized clicks are silently dropped and logged with masked user IDs. Verified by `unauthorized click is silently dropped` test in `workflow-bridge.test.ts`.

## Compatibility / Migration

- Backward compatible? **Yes** — `sendResultFooter` is an optional method on `IPlatformAdapter`; only Slack implements it. Telegram/Discord/GitHub/Web flow unchanged.
- Config/env changes? **No new env vars**. Operators must add `commands` + `reactions:write` scopes and register the two slash commands in the Slack app config and reinstall to workspace (one-time manual step, documented).
- Database migration needed? **No**.
- **Upgrade steps**:
  1. Pull `dev`, deploy
  2. In Slack app settings: add `commands` and `reactions:write` to Bot Token Scopes
  3. Toggle on Interactivity & Shortcuts (leave Request URL blank — Socket Mode handles it)
  4. Register `/archon` and `/archon-workflow` slash commands (blank Request URL)
  5. Reinstall app to workspace to issue a new token with the scopes

## Human Verification (required)

- **Verified scenarios**: Full `bun run validate` (10 packages, all tests pass). The two new unit test files cover: cost footer formatting (cost-only, tokens-only, combined, NaN/Infinity edge cases), approval/resolution/status block builders, reaction name constants, workflow_started → status post + reaction add, approval_pending → buttons in-thread, approve button → `approveWorkflow` called + message edited, reject under retry threshold → "will retry" note, cancel → `abandonWorkflow`, unauthorized click silently dropped, workflow_completed → reaction swap + cost in context block, workflow_failed → x reaction + reason.
- **Edge cases checked**: long messages now correctly footer'd with `_part i/n_`; bridge gracefully drops events for non-Slack conversations (verified by `does nothing when there is no Slack trigger` test); reactions API failures are swallowed-and-logged (won't fail the workflow); `chat.update` on a missing approval message is non-fatal (logged at warn).
- **What was *not* verified**: Live Slack workspace round-trip. Needs the new scopes installed on a real app, which has to happen after this lands. Failure modes I expect to see in real-world testing: rate-limit pushback on `chat.update` if a DAG has very fast nodes (debounced 500ms but theoretical) and `not_in_channel` errors if the bot was added to a channel without `channels:join` already there.

## Side Effects / Blast Radius (required)

- **Affected subsystems/workflows**: Slack adapter only. Other platform adapters unaffected. Orchestrator's direct-chat code paths emit an extra `sendResultFooter` call on adapters that implement it (no-op for Telegram/Discord/GitHub/Web).
- **Potential unintended effects**:
  - `chat.update` requires the bot to own the message — guaranteed since the bridge tracks `result.ts` from its own `chat.postMessage`. If the message is deleted manually, edits fail silently (logged).
  - `triggeringMessages` map is bounded at 1000 entries with FIFO eviction; chat-only conversations that never run a workflow won't grow it.
  - Stream-mode adapters now receive a `sendResultFooter` call after the final assistant chunk — currently only Slack implements it, but any adapter that adopts it should be prepared for it to run before the orchestrator decides to dispatch a workflow (current code skips it correctly on dispatch via early-return).
- **Guardrails / monitoring**: All Slack-side failures (reactions, edits, postMessage) go through `getLog().warn` / `.debug` with named events (`slack.bridge_reaction_add_failed`, `slack.bridge_status_update_failed`, etc.) — easy to grep in production logs.

## Rollback Plan (required)

- **Fast rollback**: `git revert <merge-commit>` on `dev`. The PR is self-contained — no schema changes, no migrations, no other-adapter touch points. Reverting restores the pre-PR Slack adapter behavior exactly.
- **Feature flags / toggles**: None. The bridge is unconditional once the Slack adapter is enabled. To temporarily disable: set `SLACK_BOT_TOKEN` empty (skips the whole Slack adapter init).
- **Observable failure symptoms**:
  - Slack action handlers failing → logs like `slack.bridge_approval_action_failed` with the run/node ID
  - Bolt `app.start()` failing post-attach → `slack.bot_started` log never fires; existing `slack_adapter_skipped` path triggers on missing tokens
  - Cost footer disappearing → check `orchestrator.result_footer_failed` warn logs

## Risks and Mitigations

- **Risk**: A user with `SLACK_ALLOWED_USER_IDS` access could click Cancel on another teammate's workflow.
  - **Mitigation**: Acceptable for v1 — same trust model as the existing thread-shared-state assumption. Per-user attribution is the documented follow-up (deferred from this PR's scope).
- **Risk**: Slack `chat.update` rate limit hit on DAGs with rapid node transitions.
  - **Mitigation**: 500ms trailing-edge debounce in the bridge; failures are warn-logged not thrown.
- **Risk**: Block Kit action handler must `ack()` within 3 seconds, and `approveWorkflow` / `rejectWorkflow` make DB writes.
  - **Mitigation**: All action handlers call `await ack()` *before* invoking operations. Operations are pure DB writes (no AI calls, no network) — fast in practice.
- **Risk**: Slash command flow posts a "seed" message (e.g. `<@U123> ran /archon-workflow list`) to create a thread root, which is more visible than a pure ephemeral response.
  - **Mitigation**: Documented in `slack.md`; chosen deliberately so the conversation has a real `ts` to thread under. An ephemeral acknowledgement is also sent to the invoker so they see "see thread for output."

## Summary by CodeRabbit

* **New Features**
  * Slash commands (/archon, /archon-workflow) to start workflows and seed threads.
  * In-thread interactive controls: Approve / Reject / Cancel with in-place resolution edits.
  * Workflow status messages with realtime node progress, reaction lifecycle, and tracked thread roots for later edits/reactions.
  * Long-message handling now appends per-chunk “_part i/n_” labels and reserves space for them.
  * Automatic posting of an italic cost/token footer (when available), best-effort and non-blocking.

* **Documentation**
  * Updated Slack setup, scopes, and In-Thread UX.

* **Tests**
  * Added unit tests for Block Kit builders, adapter behaviors, and workflow-to-Slack interactions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/coleam00/Archon/pull/1757?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/adapters/src/chat/slack/adapter.ts [hunk 1]
```diff
@@ -296,14 +383,130 @@ export class SlackAdapter implements IPlatformAdapter {
...
+    if (!isSlackUserAuthorized(actorId, this.allowedUserIds)) {
+      getLog().info(
+        { maskedUserId: `${actorId.slice(0, 4)}***`, kind },
+        'slack.slash_unauthorized'
+      );
+      await respond({
+        response_type: 'ephemeral',
+        text: 'Sorry — you are not on the allowed user list for this Archon instance.',
+      });
+      return;
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/adapters/src/chat/slack/adapter.ts:429]: Fixed in 7f88112 — slash command auth now silently rejects with masked log, matching the app_mention / message.im pattern.

File: packages/adapters/src/chat/slack/blocks.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    });
+  }
+
+  const footerParts: string[] = [];
+  if (snapshot.terminal && typeof snapshot.totalCostUsd === 'number') {
+    footerParts.push(`total cost: $${snapshot.totalCostUsd.toFixed(4)}`);
+  }
+  if (snapshot.terminal === 'failed' && snapshot.failureReason) {
+    footerParts.push(`reason: ${truncate(snapshot.failureReason, 200)}`);
+  }
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/adapters/src/chat/slack/blocks.ts]: Fixed in 7f88112 — cancelled runs now surface failureReason in the terminal footer too. Added a test for it.

File: packages/adapters/src/chat/slack/workflow-bridge.ts [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+        await this.adapter
+          .getApp()
+          .client.chat.postMessage({
+            channel: state.channel,
+            thread_ts: state.threadTs,
+            text: `:warning: Could not cancel run \`${runId}\`: ${err.message}`,
+          })
+          .catch(() => {
+            /* already logged */
+          });
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/adapters/src/chat/slack/workflow-bridge.ts]: Fixed in 7f88112 — generic user-facing message; raw error details stay in logs (slack.bridge_cancel_failed).

**Comment by Wirasm** [comment]
## PR Review Summary — 6-Agent Pass

`code-reviewer` · `docs-impact` · `pr-test-analyzer` · `silent-failure-hunter` · `type-design-analyzer` · `comment-analyzer`

### Critical Issues

| ID | Agent | Issue | Location |
|----|-------|-------|----------|
| C1 | code-reviewer | `@archon/workflows` imported but **not declared** in `packages/adapters/package.json` — violates the architecture boundary (`CLAUDE.md` says `@archon/adapters` depends on `@archon/core` only). Resolves today via workspace hoisting; breaks under stricter installs. | `packages/adapters/src/chat/slack/workflow-bridge.ts:17` + `packages/adapters/package.json` |
| C2 | code-reviewer | `workflow-bridge.test.ts` calls `mock.module('@archon/core', …)` and `mock.module('@archon/workflows/event-emitter', …)` (both irreversible per bun#7823) but is **not split into its own `bun test` invocation** — the file header comment claims it is, but the adapter `test` script runs the entire `src/chat/` tree in one process. No active pollution today, but the invariant is broken; next test that imports those modules in the same batch silently gets mocks. | `packages/adapters/package.json` test script |
| C3 | comment-analyzer | Debounce edge label is **factually wrong** in two places — comments say "trailing-edge" but the implementation is leading-edge (first event sets the timer, subsequent events in the burst are dropped; the timer reads current state when it fires). Will mislead anyone reasoning about burst batching. | `workflow-bridge.ts:60`, `workflow-bridge.ts:335` |

### Important Issues

| ID | Agent | Issue | Location |
|----|-------|-------|----------|
| I1 | code-reviewer | `slackBridge.detach()` never called in graceful shutdown — leaks the `WorkflowEventEmitter` subscription and lets a pending 500ms debounce fire `chat.update()` against a closed Bolt connection (swallowed but log-noisy). | `packages/server/src/index.ts` shutdown path (~L659-L699) |
| I2 | code-reviewer + silent-failure | `let comment: string \| undefined` is declared in `handleApprovalDecision`, passed through `applyResolutionEdit` → `buildApprovalResolutionBlocks`, but **never assigned**. The `&gt; ${comment}` quote in the resolution block is permanently unreachable. Also `rejectWorkflow(runId)` is called without the optional `reason` argument the operation accepts. | `workflow-bridge.ts:445-474` |
| I3 | silent-failure | `await this.applyResolutionEdit(...)` in `handleApprovalDecision` (and parallel in `handleCancelClick`) sits **outside** the try/catch. If `buildApprovalResolutionBlocks` throws, the action handler rejects unhandled — Bolt has no `app.error()` registered, so the user sees nothing on a button click. | `workflow-bridge.ts:467-476` |
| I4 | silent-failure | `handleCancelClick` exits silently when `state` is `undefined` (run already terminated) — no thread notification, no ephemeral, no log. User clicks Cancel and gets zero feedback. | `workflow-bridge.ts` cancel path |
| I5 | type-design | `body as ActionBody` / `action as ActionElement` casts bypass Bolt's `BlockAction`/`ButtonAction` types and depart from the SDK-types convention in CLAUDE.md ("Import SDK types directly"). Future Bolt SDK field changes won't be caught at compile time. | `workflow-bridge.ts:418-426` |
| I6 | pr-test-analyzer | **Slash-command authorization** is entirely untested (criticality 9/10). If `parseAllowedUserIds` ever regresses (e.g. empty-string user always authorized), no test catches it. | `adapter.ts:237-247` |
| I7 | pr-test-analyzer | **`triggeringMessages` 1000-entry FIFO eviction** is untested (criticality 8/10). Also: `Map.set()` on an existing key does NOT move it to the end — re-tracked conversations stay at their old FIFO position. Logic is correct (verified) but the eviction trigger and ordering deserve a test. | `adapter.ts:209-216` |
| I8 | pr-test-analyzer | **Slash-command seed-post failure path** untested (criticality 8/10) — if `chat.postMessage` throws, the ephemeral error path and the "do-not-call-messageHandler-with-undefined-ts" guard are unverified. | `adapter.ts:283-293` |
| I9 | pr-test-analyzer | **Multi-chunk `_part i/n_` annotation boundary** untested (criticality 6/10) — single-chunk path (`total === 1` skips annotation) is easy to regress. | `adapter.ts:88-94` |
| I10 | pr-test-analyzer | **`rejectWorkflow → { cancelled: true, maxAttemptsReached: true }` branch** untested (criticality 7/10). Three distinct outcome strings, only the "will retry" branch is asserted. | `workflow-bridge.ts:455-458` |
| D1 | docs-impact | `packages/docs-web/src/content/docs/reference/architecture.md:102-107` IPlatformAdapter listing does not include the new `sendResultFooter?` method. | architecture.md |
| D2 | docs-impact | `packages/docs-web/src/content/docs/guides/approval-nodes.md:57-60` says "Natural-language messages, the CLI, and the Web UI approve button all auto-resume" — now also true for Slack in-thread Approve/Reject. | approval-nodes.md |
| D3 | docs-impact | `CLAUDE.md:133` test-isolation section says `@archon/adapters (3)` but the package has 5 batches (pre-existing). | CLAUDE.md |
| Cmt1 | comment-analyzer | `removeReactionSafe` silent catch has no rationale comment — the parallel `addReactionSafe` documents `already_reacted`/rate-limit are non-fatal; the remove counterpart should explain `no_reaction`/`already_removed` similarly (CLAUDE.md mandates documenting intentional fallbacks). | `workflow-bridge.ts:393-406` |
| Cmt2 | comment-analyzer | `STATUS_UPDATE_DEBOUNCE_MS = 500` has no rationale — Slack `chat.update` rate limit (~1/s/channel) is the non-obvious driver; future tuners have no anchor. | `workflow-bridge.ts:66` |

### Suggestions

| ID | Agent | Suggestion | Location |
|----|-------|------------|----------|
| S1 | type-design | Use `TokenUsage` from `@archon/providers/types` in `IPlatformAdapter.sendResultFooter` instead of the inline `{ input; output; total?; cost? }` — already imported via `MessageChunk`. | `packages/core/src/types/index.ts:175` |
| S2 | comment-analyzer | Drop the `"for v1"` tag in `// Loop / tool / artifact events are surface-noisy — skipped for v1.` Either keep the rationale alone or convert to a tracked TODO. | `workflow-bridge.ts:156` |
| S3 | comment-analyzer | Remove what-comments: `// Slash commands: /archon (general) and /archon-workflow (workflow control)` (adapter.ts:394); the second sentence of the `blocks.ts` file-level doc; `blocks.test.ts` file header. | various |

### Strengths

- `handleEvent` top-level try/catch with `slack.bridge_event_handler_failed` event name + structured context — exactly the pattern CLAUDE.md prescribes.
- `addReactionSafe` correctly documents why swallowing is intentional.
- `maybeSendResultFooter` in `orchestrator-agent.ts:1431-1444` is properly wrapped, logs `orchestrator.result_footer_failed` at `.warn` with structured context, and guards on method presence.
- `handleApprovalDecision` DB-error path falls through to a user-visible `outcomeNote = 'error: …'` in the resolution message instead of silently failing.
- `unauthorized click is silently dropped` test asserts the mock operation is not called rather than peeking at log output — correct behavioral verification.
- `formatCostFooter` comprehensively tested for NaN/Infinity/zero-token/million-scale boundaries.
- File-header lifecycle ordering comment (`attach() BEFORE SlackAdapter.start()`) captures a non-obvious temporal invariant.
- `_part i/n_` reserve-headroom comment (adapter.ts:85-87) is a textbook WHY-comment with a worked example.
- `KnownBlock` from `@slack/bolt` is used directly per the SDK-types convention.
- `NodeState` / `RunTerminalState` discriminated unions + `Record<NodeState, string>` glyph maps give compile-time exhaustiveness.

### Verdict: NEEDS FIXES

C1–C3 are blocking. I1–I5 are real correctness/UX gaps that should land before merge; the test gaps (I6–I10) and docs (D1–D3) can land alongside or in a fast follow.

### Recommended Order

1. Fix C1 by adding `"@archon/workflows": "workspace:*"` to `packages/adapters/package.json`.
2. Fix C2 by splitting `workflow-bridge.test.ts` into its own `bun test` invocation in the adapter `test` script.
3. Fix C3 by either changing the comments to "leading-edge" or adding the rate-limit rationale (Cmt2 in the same edit).
4. Address I1 (wire `slackBridge.detach()` into `shutdown()`), I2 (delete dead `comment` plumbing or wire it up), I3 (widen the try/catch), I4 (log + ephemeral on cancel-with-no-state).
5. Add tests for I6–I10.
6. Doc fixes D1–D3.
7. Polish: I5 (Bolt types), S1 (TokenUsage), Cmt1, Cmt2, S2, S3.
